Fold node-stats telemetry device fields in one doc per sample

dliappis opened the pull request:
This is similar to the work done earlier for the ccr-stats telemetry
device in 198ceefc93d223cdb292fb15b1373b64058f28cb

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
-                                          include=["docs", "store", "indexing", "search", "merges", "query_cache", "fielddata",
-                                                   "segments", "translog", "request_cache"])
+                collected_node_stats.update(
+                    self.collect_indices_stats(node_name, node_stats,
```
Maybe we could benefit from shorter names by removing the `collect_` prefix from all methods? To some extent these methods are more like extractors or getters so I think it would be ok here.

danielmitterdorfer commented on tests/mechanic/telemetry_test.py:
```diff
@@ -418,213 +418,324 @@ def test_stores_filtered_ccr_stats(self, metrics_store_put_doc):
+        # https://stackoverflow.com/a/28418085/108668
+        call = metrics_store_put_doc.call_args
+        call_args, _ = call
+        self.maxDiff = None
```
maybe set this rather in the constructor or in `setUp` as this has side-effects on other methods?

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
+
+        :param prefix: The prefix for all flattened values. _ will be used as a separator. Defaults to None.
+        :param stats: Dict with values to be flattened, using _ as a separator. Defaults to {}.
+        :param include: List of keys to process in stats. Filtering will be done only for top level keys.
```
I think this is only necessary for the `indices_stats` section? Can the caller do this instead? This would simplify this method a bit.

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
         time.sleep(self.sample_interval)
 
-    def record_indices_stats(self, node_name, node_stats, include):
+    def flatten_stats_fields(self, prefix=None, stats=None, include=None, level=0):
```
IMHO this method is pretty complex and does more than just flattening fields:

1. It flattens
2. It filters top-level keys
3. It formats the flattened keys

Can we separate those responsibilities so it is simpler to understand?

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
+                            if isinstance(subsection_value, (int, float)) and not isinstance(subsection_value, bool):
+                                yield "{}{}_{}".format(prefix + "_" if prefix else "", section_name, subsection_name), subsection_value
+                    else:
+                        if isinstance(section_value, (int, float)) and not isinstance(section_value, bool):
```
I wonder whether we can get rid of this code duplication when simplifying this method as suggested above.

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
+        :param stats: Dict with values to be flattened, using _ as a separator. Defaults to {}.
+        :param include: List of keys to process in stats. Filtering will be done only for top level keys.
+        :param level: Int used to control filtering at the top level; only used internally. Defaults to 0.
+        :return:
```
Missing documentation for return value?

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,107 +481,90 @@ def record(self):
+        :param prefix: The prefix for all flattened values. _ will be used as a separator. Defaults to None.
+        :param stats: Dict with values to be flattened, using _ as a separator. Defaults to {}.
+        :param include: List of keys to process in stats. Filtering will be done only for top level keys.
+        :param level: Int used to control filtering at the top level; only used internally. Defaults to 0.
```
We could have an internal function that defines this parameter and hide it from the top-level function?

danielmitterdorfer commented on tests/mechanic/telemetry_test.py:
```diff
@@ -418,213 +418,324 @@ def test_stores_filtered_ccr_stats(self, metrics_store_put_doc):
-            mock.call(node_name="rally0", name="jvm_buffer_pool_mapped_count", count=7, meta_data=metrics_store_meta_data),
-            mock.call(node_name="rally0", name="jvm_buffer_pool_direct_count", count=6, meta_data=metrics_store_meta_data),
-        ], any_order=True)
+        # Extract mock call args, needed to perform dict comparison of first param of mocked method
```
That's really too bad that we have to do this. Would it help if we use a sorted dict? If not, maybe define a helper method named e.g. `assertNodeStatsRecorded()` that encapsulates this (you also need this in the next test)?

dliappis commented:
@danielmitterdorfer Thanks for the review. I addressed all the comments, apart from taking out the formatting from the flatten method (as it's now much simpler and more readable). Would appreciate another review iteration.

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
-        thread_pool_stats = node_stats["thread_pool"]
-        return self.flatten_stats_fields(prefix="thread_pool", stats=thread_pool_stats)
+    def thread_pool_stats(self, node_name, node_stats):
+        thr_pool_stats = node_stats["thread_pool"]
```
Maybe just extract inline? I.e.: 

```
return self.flatten_stats_fields(prefix="thread_pool", stats=node_stats["thread_pool"])
```

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
 
-    def collect_circuit_breaker_stats(self, node_name, node_stats):
+    def circuit_breaker_stats(self, node_name, node_stats):
         breaker_stats = node_stats["breakers"]
```
inline similar to `thread_pool_stats`

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
 
-    def collect_jvm_buffer_pool_stats(self, node_name, node_stats):
+    def jvm_buffer_pool_stats(self, node_name, node_stats):
         buffer_pool_stats = node_stats["jvm"]["buffer_pools"]
```
inline similar to `thread_pool_stats`

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
 
-    def collect_jvm_mem_stats(self, node_name, node_stats):
+    def jvm_mem_stats(self, node_name, node_stats):
         mem_stats = node_stats["jvm"]["mem"]
```
inline similar to `thread_pool_stats`

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
 
-    def collect_network_stats(self, node_name, node_stats):
+    def network_stats(self, node_name, node_stats):
         transport_stats = node_stats.get("transport")
```
inline similar to `thread_pool_stats`

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -481,90 +482,90 @@ def record(self):
-        process_stats = node_stats["process"]["cpu"]
-        return self.flatten_stats_fields(prefix="process_cpu", stats=process_stats)
+    def process_stats(self, node_name, node_stats):
+        proc_stats = node_stats["process"]["cpu"]
```
inline similar to `thread_pool_stats`
